muon/_core/preproc.py

```diff
@@ -155,9 +155,25 @@ def _l2norm(
     adata: AnnData, rep: Optional[Union[Iterable[str], str]] = None, n_pcs: Optional[int] = 0
 ):
     X = _choose_representation(adata, rep, n_pcs)
-    norm = X / np.linalg.norm(X, ord=2, axis=1, keepdims=True)
-    norm[~np.isfinite(norm)] = 0
+    sparse_X = issparse(X)
+    if sparse_X:
+        X_norm = linalg.norm(X, ord=2, axis=1)
+        norm = X / np.expand_dims(X_norm, axis=1)
+        if issparse(norm):
+            # find nan and infinite values and construct matrix without
+            i, j, val = find(norm)
+            isfin = np.isfinite(val)
+            i = i[isfin]
+            j = j[isfin]
+            val = val[isfin]
+            norm = csr_matrix((val, (i, j)), shape=X.shape)
```

Comment:
I think for `csr_matrix`, `csc_matrix`, and `coo_matrix` it would probably be  more efficient to do something like `norm.data[~np.isfinite(norm.data)] = 0`. Sparse matrices in AnnData are usually CSR or CSC (in fact, IO only has support for CSR and CSC implemented), so it would probably make sense to also add that special case.

Reply:
Thats a very good point! I simplified it in commit https://github.com/scverse/muon/pull/114/commits/e7b7b12ace6f77086c06aa1194a8159060004645